spec/unit/lib/performance_platform/gateway/completion_rate_spec.rb

```diff
@@ -0,0 +1,94 @@
+describe PerformancePlatform::Gateway::CompletionRate do
+  let(:user_repo) { WifiUser::Repository::User }
+
+  before do
+    DB[:userdetails].truncate
+
+    # Outside of date scope
+    user_repo.create(
+      username: '1',
+      created_at: Date.today - 1,
+      contact: '+1234567890',
+      sponsor: '+1234567890'
+    )
+
+    # Outside of date scope
+    # and logged in
+    user_repo.create(
+      username: '2',
+      created_at: Date.today - 1,
+      contact: '+1234567890',
+      sponsor: '+1234567890',
+      last_login: Date.today
+    )
+
+    # SMS self-registered within date scope
+    user_repo.create(
+      username: '3',
+      created_at: Date.today - 8,
+      contact: '+2345678901',
+      sponsor: '+2345678901'
+    )
+
+    # SMS self-registered within date scope
+    # and logged in
+    user_repo.create(
+      username: '4',
+      created_at: Date.today - 8,
+      contact: '+2345678901',
+      sponsor: '+2345678901',
+      last_login: Date.today
+    )
+
+    # SMS sponsor-registered within date scope
+    user_repo.create(
+      username: '5',
+      created_at: Date.today - 8,
+      contact: '+2345678901',
+      sponsor: 'sponsor@example.com'
+    )
+
+    # SMS sponsor-registered within date scope
+    # and logged in
+    user_repo.create(
+      username: '6',
+      created_at: Date.today - 8,
+      contact: '+2345678901',
+      sponsor: 'sponsor@example.com',
+      last_login: Date.today
+    )
+
+    # email self-registered within scope
+    user_repo.create(
+      username: '7',
+      created_at: Date.today - 10,
+      contact: 'me@example.com',
+      sponsor: 'me@example.com'
+    )
+
+    # Email self-registered within scope
+    # and logged in
+    user_repo.create(
+      username: '8',
+      created_at: Date.today - 10,
+      contact: 'me@example.com',
+      sponsor: 'me@example.com',
+      last_login: Date.today
+    )
+  end
+
+  context 'given no signups' do
+    it 'returns stats with zero signups' do
```

Comment:
Are these stats below for 0 signups?
Are sms_registered and email_registered signups?

Reply:
Mhm, no, just a copy left over, good catch 👍